apollo-router/src/query_planner/dual_query_planner.rs

```diff
@@ -385,26 +372,46 @@ fn flatten_node_matches(this: &FlattenNode, other: &FlattenNode) -> bool {
 // AST comparison functions
 
 fn same_ast_document(x: &ast::Document, y: &ast::Document) -> bool {
-    x.definitions
-        .iter()
-        .zip(y.definitions.iter())
-        .all(|(x_def, y_def)| same_ast_definition(x_def, y_def))
-}
-
-fn same_ast_definition(x: &ast::Definition, y: &ast::Definition) -> bool {
-    match (x, y) {
-        (ast::Definition::OperationDefinition(x), ast::Definition::OperationDefinition(y)) => {
-            same_ast_operation_definition(x, y)
+    fn split_definitions(
+        doc: &ast::Document,
+    ) -> (
+        Vec<&ast::OperationDefinition>,
+        Vec<&ast::FragmentDefinition>,
+        Vec<&ast::Definition>,
+    ) {
+        let mut operations: Vec<&ast::OperationDefinition> = Vec::new();
+        let mut fragments: Vec<&ast::FragmentDefinition> = Vec::new();
+        let mut others: Vec<&ast::Definition> = Vec::new();
+        for def in doc.definitions.iter() {
+            match def {
+                ast::Definition::OperationDefinition(op) => operations.push(op),
+                ast::Definition::FragmentDefinition(frag) => fragments.push(frag),
+                _ => others.push(def),
+            }
         }
-        (ast::Definition::FragmentDefinition(x), ast::Definition::FragmentDefinition(y)) => x == y,
-        _ => false,
+        fragments.sort_by_key(|frag| frag.name.clone());
+        (operations, fragments, others)
     }
-}
 
-fn hash_value<T: Hash>(x: &T) -> u64 {
-    let mut hasher = DefaultHasher::new();
-    x.hash(&mut hasher);
-    hasher.finish()
+    let (x_ops, x_frags, x_others) = split_definitions(x);
+    let (y_ops, y_frags, y_others) = split_definitions(y);
+
+    assert!(x_others.is_empty(), "Unexpected definition types");
+    assert!(y_others.is_empty(), "Unexpected definition types");
+    assert!(
+        x_ops.len() == y_ops.len(),
+        "Different number of operation definitions"
+    );
+
+    x_ops
+        .iter()
+        .zip(y_ops.iter())
+        .all(|(x_op, y_op)| same_ast_operation_definition(x_op, y_op))
+        && x_frags.len() == y_frags.len()
```

Comment:
Is there a reason that we're panicking if `x_ops.len()` doesn't equal `y_ops.len()` but not for the fragments?

Reply:
The number of operations shouldn't really different at all (just one op all the time). So, I thought it was assertion-worthy.

Reply:
I'm rethinking this. It's running in production. It shouldn't be using assert. I'll change that.